src/anbox/qemu/sensors_message_processor.cpp

```diff
@@ -16,28 +16,69 @@
  */
 
 #include "anbox/qemu/sensors_message_processor.h"
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/lexical_cast.hpp>
+#include <chrono>
+#include <thread>
+
 #include "anbox/logger.h"
 
+using namespace std;
+
 namespace anbox {
 namespace qemu {
 SensorsMessageProcessor::SensorsMessageProcessor(
-    const std::shared_ptr<network::SocketMessenger> &messenger)
-    : QemudMessageProcessor(messenger) {}
+    shared_ptr<network::SocketMessenger> messenger, shared_ptr<application::SensorsState> sensorsState)
+    : QemudMessageProcessor(messenger), sensorsState_(sensorsState) {
+  thread = std::thread([this]() {
+    for (;;) {
+      if (temperature.load())
+        send_message("temperature:" + to_string(sensorsState_->temperature));
+      if (!run_thread.load())
+        break;
+      this_thread::sleep_for(delay.load() * 1ms);
+    }
+  });
+}
+
+SensorsMessageProcessor::~SensorsMessageProcessor() {
+  run_thread = false;
+  thread.join();
+}
 
-SensorsMessageProcessor::~SensorsMessageProcessor() {}
+void SensorsMessageProcessor::handle_command(const string &command) {
+  if (!(list_sensors(command) || handle_set(command) || handle_set_delay(command)))
+    ERROR("Unknown command: " + command);
+}
+
+bool SensorsMessageProcessor::handle_set_delay(const string &command) {
+  if (!boost::starts_with(command, "set-delay:"))
+    return false;
+  delay = boost::lexical_cast<int>(command.substr("set-delay:"s.length()));
```

Comment:
to be honest, maybe I'm used to C, but I find this much more clear and readable:
```C
if (!sscanf(command, "set-delay:%i", &delay))
    error();
```

compare: "scan string "command", looking for "set-delay:" followed by an int, and put me the int in $delay."

and the alternative "delay is the result of the conversion to an int of the $command string but where you dropped the $n first int, where n is what you get when you take "set-delay" as a string and then count the characters"

Reply:
You might be right... Not sure how would sscanf work with `std::atomic` though. How about solution with Boost Sprit X3?

Reply:
My personal opinion on the question is that the code was actually more readable before. In 10 minutes, I could not grasp what x3 really is, or how the new code actually works (like, I get what it's supposed to do, but I'd be quite embarrassed if I had to debug it). I think if you've not used x3 or C++ in a professional setting, this version is harder to read.